File: mt-user/src/main/kotlin/com/mt/mtuser/common/page/PageQuery.java
package com.mt.mtuser.common.page;

import org.hibernate.validator.constraints.Range;
import org.springframework.util.StringUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

/**
 * Created by gyh on 2019/1/24.
 *
 * @apiDefine PageQuery
 * @apiParam {String} [searchField] 查找字段,如'name'
 * @apiParam {String} [searchOper] 查找方式,可用参数与说明: [{'cn':包含'},{'eq':'等于'},{'nc':'不包含'},{'ne':'不等于'},{'gt':'大于'},{'gt;=':'大于等于'},{'lt':'小于'},{'lt;=':'小于等于'},{'bw':'开始于'},{'bn':'不开始于'},{'ew':'结束于'},{'en':'不结束于'}]
 * @apiParam {String} [searchString] 查找的字符,如'张'
 * @apiParam {Int} [pageNum] 当前页号,页从1开始,默认第一页
 * @apiParam {Int} [pageSize] 每页显示的数量,默认30条
 * @apiParam {String} [orderBy] 排序规则,如: create_time DESC,update_time ASC
 */
public class PageQuery {
    private static final Logger loggger = Logger.getLogger(PageQuery.class.getSimpleName());
    /**
     * 按id递增排序
     */
    public static final String ORDER_BY_CREATE_TIME_DESC = "id ASC";

    private static final Map<String, String> SearchType = new HashMap<>();
    private static final String[][] nmka = {
            {"cn", "like", "%%%s%%"}, {"eq", "=", "%s"},
            {"nc", "not like", "%%%s%%"}, {"ne", "<>", "%s"},
            {"gt", ">", "%s"}, {"lt", "<", "%s"},
            {"bw", "like", "%s%%"}, {"bn", "not like", "%s%%"},
            {"ew", "like", "%%%s"}, {"en", "not like", "%%%s"}
    };
    // "查找字段", example = "name")
    private String searchField;
    // "查找方式", allowableValues = "[['cn', '包含'], ['eq', '等于'], ['nc', '不包含'], ['ne', '不等于'], ['gt', '大于'], ['lt', '小于'], ['bw', '开始于'], ['bn', '不开始于'], ['ew', '结束于'], ['en', '不结束于']]")
    private String searchOper;
    // "查找的字符", example = "张")
    private String searchString;

    // "当前页号")
    @Range(min = 1, max = Integer.MAX_VALUE)
    private int pageNum = 1;

    // "一页数量")
    @Range(min = 1, max = Integer.MAX_VALUE)
    private int pageSize = 30;

    // "排序", notes = "例：create_time desc,update_time ASC")
    private String orderBy;

    public PageQuery(int pageNum, int pageSize) {
        this.pageNum = pageNum;
        this.pageSize = pageSize;
    }

    public PageQuery() {
    }

    public int getOffset() {
        return (this.pageNum - 1) * this.pageSize;
    }

    private String buildSubSql() {
        StringBuilder sql = new StringBuilder();
        if (!StringUtils.isEmpty(searchField) && !StringUtils.isEmpty(searchOper) && !StringUtils.isEmpty(searchString)) {
            for (String[] strings : nmka) { //判断指令是否在允许范围内
                if (strings[0].equals(searchOper)) {
                    return sql
                            .append(" ")
                            .append(searchField).append(" ")
                            .append(strings[1]).append(" '")
                            .append(String.format(strings[2], searchString)).append("'")
                            .toString();
                }
            }
        }
        loggger.info(sql.toString());
        return sql.toString();
    }

    private String buildLimitSql() {
        return " limit " +
                pageSize +
                " OFFSET " +
                pageSize * (pageNum - 1);
    }

    public QueryBuild where() {
        return new QueryBuild(this).where();
    }

    public QueryBuild query() {
        return new QueryBuild(this).where();
    }

    public QueryBuild limit() {
        return new QueryBuild(this).limit();
    }

    public static class QueryBuild {
        private String where = " ";
        private String query = " ";
        private String limit = " ";
        private PageQuery pageQuery;

        public QueryBuild(PageQuery pageQuery) {
            this.pageQuery = pageQuery;
        }

        public QueryBuild where() {
            where = " where";
            return this;
        }

        public QueryBuild query() {
            query = pageQuery.buildSubSql();
            return this;
        }

        public QueryBuild limit() {
            limit = pageQuery.buildLimitSql();
            return this;
        }

        public String build() {
            String sql = StringUtils.isEmpty(query.trim())
                    ? query + limit
                    : where + query + limit;
            loggger.info(sql);
            return sql;
        }
    }

    public Integer getPageNum() {
        return pageNum;
    }

    public void setPageNum(int pageNum) {
        this.pageNum = pageNum;
    }

    public Integer getPageSize() {
        return pageSize;
    }

    public void setPageSize(int pageSize) {
        this.pageSize = pageSize;
    }

    public String getOrderBy() {
        return orderBy;
    }

    public void setOrderBy(String orderBy) {
        this.orderBy = orderBy;
    }

    public String getSearchField() {
        return searchField;
    }

    public void setSearchField(String searchField) {
        this.searchField = searchField;
    }

    public String getSearchOper() {
        return searchOper;
    }

    public void setSearchOper(String searchOper) {
        this.searchOper = searchOper;
    }

    public String getSearchString() {
        return searchString;
    }

    public void setSearchString(String searchString) {
        this.searchString = searchString;
    }


}

File: mt-user/src/main/kotlin/com/mt/mtuser/entity/page/PageQuery.kt
package com.mt.mtuser.entity.page

import com.mt.mtcommon.exception.BusinessException
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.relational.core.query.Criteria

/**
 * Created by gyh on 2020/3/20.
 * @apiDefine PageQuery
 * @apiParam {String} [searchField] 查找字段,如'name'
 * @apiParam {String} [searchOper] 查找方式,可用参数与说明: [{'cn':包含'},{'eq':'等于'},{'ne':'不等于'},{'gt':'大于'},
 * {'gt;=':'大于等于'},{'lt':'小于'},{'lt;=':'小于等于'},{'bw':'开始于'},{'ew':'结束于'}]
 * @apiParam {String} [searchString] 查找的字符,如'张'
 * @apiParam {Int} [pageNum] 当前页号,页从0开始,默认第0页
 * @apiParam {Int} [pageSize] 每页显示的数量,默认30条
 * @apiParam {String} [order] 排序字段,如: id
 * @apiParam {String} [direction] 排序规则,如: DESC, ASC
 */
class PageQuery(
    val pageNum: Int = 0,
    val pageSize: Int = 30,
    val searchField: String? = null,    // 查找字段 :name
    val searchOper: String? = null,     // 查找方式 :"[['cn', '包含'], ['eq', '等于'], ['nc', '不包含'], ['ne', '不等于'], ['gt', '大于'], ['lt', '小于'], ['bw', '开始于'], ['bn', '不开始于'], ['ew', '结束于'], ['en', '不结束于']]")
    val searchString: String? = null,   // 查找的字符 :张
    var order: String? = null,          // 排序字段 :id
    var direction: String? = null       // 排序方向 :asc
) {
    private val logger = LoggerFactory.getLogger(this.javaClass)
    private val oper = arrayOf(
        arrayOf("cn", "like", "%%%s%%"), arrayOf("eq", "=", "%s"),
        arrayOf("nc", "not like", "%%%s%%"), arrayOf("ne", "<>", "%s"), arrayOf("gt", ">", "%s"),
        arrayOf("lt", "<", "%s"), arrayOf("bw", "like", "%s%%"), arrayOf("bn", "not like", "%s%%"),
        arrayOf("ew", "like", "%%%s"), arrayOf("en", "not like", "%%%s")
    )

    fun where(alias: String = ""): Criteria {
        var criteria = Criteria.empty()
        if (searchField != null && searchOper != null && searchString != null) {
            val column = if (alias.isBlank()) searchField else "$alias.$searchField"
            with(Criteria.where(column)) {
                criteria = when (searchOper) {
                    "eq" -> `is`(searchString.toIntOrNull() ?: searchString)
                    "ne" -> not(searchString.toIntOrNull() ?: searchString)
                    "gt" -> greaterThan(searchString.toIntOrNull() ?: searchString)
                    "gt;=" -> greaterThanOrEquals(searchString.toIntOrNull() ?: searchString)
                    "lt" -> lessThan(searchString.toIntOrNull() ?: searchString)
                    "lt;=" -> lessThanOrEquals(searchString.toIntOrNull() ?: searchString)
                    "cn", "bw", "ew" -> buildSubSql(alias)
                    "nc", "bn", "en" -> throw BusinessException("暂不支持的查找方式$searchOper")
                    else -> throw BusinessException("不支持的查找方式$searchOper")
                }
            }
        }
        return criteria
    }

    fun page(): Pageable {
        var pageable = PageRequest.of(pageNum, pageSize)
        if (order != null) {
            pageable = PageRequest.of(
                pageNum,
                pageSize,
                Sort.by(Sort.Direction.fromString(direction ?: Sort.Direction.ASC.name), order)
            )
        }
        return pageable
    }

    fun toPageSql(): String {
        val sb = StringBuilder()
        if (order != null) {
            if (direction.equals("asc", true) || direction.equals("desc", true)) {
                sb.append(" order by ").append(order).append(" ").append(direction).append(" ")
            }
        }
        if (pageNum == 0) {
            sb.append(" limit ").append(pageSize)
        } else {
            val offset = pageNum * pageSize
            sb.append(" limit ").append(pageSize).append(" OFFSET ").append(offset).append(" ")
        }
        return sb.toString()
    }

    private fun buildSubSql(alias: String = ""): Criteria {
        if (searchField != null && searchOper != null && searchString != null) {
            for (strings in oper) { //判断指令是否在允许范围内
                if (strings[0] == searchOper) {
                    val column = if (alias.isBlank()) searchField else "$alias.$searchField"
                    return Criteria.where(column).like(String.format(strings[2], searchString))
                }
            }
        }
        return Criteria.empty()
    }

    override fun toString(): String {
        return "PageQuery(pageNum=$pageNum, pageSize=$pageSize, searchField=$searchField, searchOper=$searchOper, searchString=$searchString, order=$order, direction=$direction, logger=$logger, oper=${oper.contentToString()})"
    }

}